### PR TITLE
support for --relaxed-rpc in anvil

### PR DESCRIPTION
**Pull request by aathan**
## Motivation

Sometimes you have no choice but to deal with an upstream rpc provider that does not strictly conform to the standards. It would be nice to be able to tell anvil to be a little more forgiving in those scenarios.

## Solution

Ultimately, this means having Providers that can vary behavior based on parameters. I've submitted an upstream patch to ethers-rs to support this (with an initial implementation that has a single const bool generic type param for "strictness") here: https://github.com/gakonst/ethers-rs/pull/2527, which is necessary to compile this patch.

### CODE REVIEW

**Comment by aathan** [comment]
I'm new to Rust. There may be better ways to achieve the behavior polymorphism. I didn't want to use a dyn trait because that would introduce dynamic dispatch carelessly, and I got compiler warnings that impl traits are unstable, so this uses a ProviderTypes enum.

File: common/src/provider.rs [hunk 1]
```diff
@@ -1,18 +1,48 @@
...
 };
 use eyre::WrapErr;
 use reqwest::{IntoUrl, Url};
-use std::{borrow::Cow, time::Duration};
+use serde::{de::DeserializeOwned, Serialize};
+use std::{borrow::Cow, fmt::Debug, time::Duration};
+
+/// Wraps strict or loose rpc handling in the Http provider
+#[derive(Debug)]
+pub enum ProviderTypes {
```
Comment (mattsse) [inline common/src/provider.rs:18]: would like to rename this
```suggestion
pub enum ProviderKind {
```

File: common/src/provider.rs [hunk 2]
```diff
@@ -1,18 +1,48 @@
...
+    is_local_endpoint, Http, HttpRateLimitRetryPolicy, JsonRpcClient, LooseHttp, Middleware,
+    Provider, RetryClient, RetryClientBuilder, DEFAULT_LOCAL_POLL_INTERVAL,
 };
 use eyre::WrapErr;
 use reqwest::{IntoUrl, Url};
-use std::{borrow::Cow, time::Duration};
+use serde::{de::DeserializeOwned, Serialize};
+use std::{borrow::Cow, fmt::Debug, time::Duration};
+
+/// Wraps strict or loose rpc handling in the Http provider
```
Comment (mattsse) [inline common/src/provider.rs:16]: I'd appreciate a few more docs here, why we're doing this

**Review by mattsse — CHANGES_REQUESTED** [review]
this doesn't look too bad.

supportive, because this comes up regularly, especially on new testnet nodes.

a few nits

File: anvil/src/cmd.rs [hunk 3]
```diff
@@ -411,6 +412,12 @@ pub struct AnvilEvmArgs {
     #[clap(long, requires = "fork_url", help_heading = "Fork config")]
     pub no_storage_caching: bool,
 
+    /// Specify whether to be strict in json rpc handling of the upstream provider
```
Comment (Evalir) [inline anvil/src/cmd.rs:415]: This docs are good—let's add just a tad more context as to _why_ we're doing this, talk a little bit about hardhat's node handling and such.

File: common/src/provider.rs [hunk 4]
```diff
@@ -1,18 +1,48 @@
...
+#[derive(Debug)]
+pub enum ProviderTypes {
+    ///strict
+    Http(Http),
+    ///loose
+    LooseHttp(LooseHttp),
```
Comment (Evalir) [inline common/src/provider.rs:22]: I think i'd also rename `LooseHttp` to `RelaxedHttp` to be consistent with all the other naming

File: common/src/provider.rs [hunk 5]
```diff
@@ -1,18 +1,48 @@
...
+        T: Debug + Serialize + Send + Sync,
+        R: DeserializeOwned + Send,
+    {
+        match self {
+            ProviderTypes::Http(i) => i.request(method, params).await,
+            ProviderTypes::LooseHttp(i) => i.request(method, params).await,
+        }
+    }
+}
+//.await.map_err(Self::Error::into)?,
```
Comment (Evalir) [inline common/src/provider.rs:42]: should we delete this?

File: common/src/provider.rs (hunk 2 above)
Comment (Evalir) [reply common/src/provider.rs:16]: + 1, let's be verbose with docs as it's not obvious from the get go

File: anvil/src/config.rs [hunk 6]
```diff
@@ -693,6 +696,14 @@ impl NodeConfig {
         self
     }
 
+    /// Sets whether to process rpc in a relaxed way (for now this just means if method: appears in
+    /// the response, ignore it)
```
Comment (Evalir) [inline anvil/src/config.rs:700]:
```suggestion
    /// Sets whether to process rpc in a relaxed way, which allows the usage of rpc providers that loosely follow the JSON-RPC spec
```

**Review by Evalir — CHANGES_REQUESTED** [review]
some comments, supportive to have parity with hardhat's support of these non-compliant rpc providers

**Comment by aathan** [comment]
I've implemented your suggestions. If you want more changes to the docs/comments please do so as you please ... note that I kept the original comments vague in most locations "downstream" of clap precisely because this option is likely to become vague down the road. Right now it only affects whether "method" can appear in the rpc json response dict, but in the future it might be 10 other things. Do we really want to update every comment everywhere every time there is a new "relaxation"? Maybe. Up to you...

**Comment by aathan** [comment]
Also, I found other places where providers are instantiated (like if I remember correctly, where one of the eth_ methods that sets the upstream rpc provider is handled). In general it seems like anvil is not terribly consistent in how it instantiates things (I've commented on this in one of my bug reports, if you're curious about another example I can dig that up) -- it hasn't always been clear to me where various cfg structs should be significant and/or passed around and/or consulted, but anvil may need a cleanup pass relative to this issue. Specific to relaxed-rpc, if you specified this on the command line, shouldn't it apply everywhere?

For example, I saw but did not dig into response parsers relative to websockets? That should get this feature too, no?

Finally, I want to again highlight that I'm totally new to Rust. My use of a `ProviderKinds` variant is just my noob attempt at behavior polymorphism. It feels like we should be able to instead carry around a `impl Trait` reference, such that (I would assume) the compiler can know at compile time to create a dispatch to the necessary deserializer without a dynamic lookup (just a function pointer at the lowest level)... but I'm not familiar enough with Rust to reason on this issue.

Please review that part of the design with this in mind.

**Comment by aathan** [comment]
Do I need to refactor the ethers-rs feature into a non-generic runtime-bool to make this feature happen? I left a comment over on that project.

Meanwhile, I brought this PR up to date, hopefully the changes to ethers-rs can be approved and then this can be merged.

This is an imperfect PR that does not address some bigger problems I see in the codebase. Specifically, that remote addresses (urls, ips, whatever) are stored basically bare, and then, providers are instantiated from them.

When those new providers are created, there's no attempt to mimic the connectivity features relevant to successfully communicating with that far end. Timeouts, relaxed/strict rpc handling, ssl certs, whatever.

Every place that stores a "fork_url" w/ or w/o separate connectivity parameters should probably instead store a "RemoteConnInfo" struct that either contains or references enough info from which to build connectivity providers. I did notice there are both ForkUrl and RpcUrl types floating around.

**Comment by Evalir** [comment]
Will also close this as stale, mainly because we are now knee deep in the larger ecosystem migration to Alloy, and eventually we'll move to replacing provider code.

We should set as an objective to support this flag—albeit doing more work on this PR to bring it to a mergeable state is probably not worth it anymore as it will soon be changed.

**Pull request closed without merging**